File: reverie/modes.py
"""
Mode registry for Reverie.

Centralizes aliases, descriptions, and mode-switching rules so the CLI,
tooling layer, and system prompts stay in sync.
"""

from __future__ import annotations

from typing import Any, Dict, List


MODE_METADATA: Dict[str, Dict[str, object]] = {
    "reverie": {
        "display_name": "Reverie",
        "description": "General-purpose coding mode that uses Context Engine plus the smallest useful toolset for focused fixes and end-to-end delivery.",
        "switchable": True,
    },
    "reverie-atlas": {
        "display_name": "Reverie-Atlas",
        "description": "Document-driven spec development mode for complex systems, pairing deep research with Context Engine and Atlas delivery artifacts.",
        "switchable": True,
    },
    "reverie-gamer": {
        "display_name": "Reverie-Gamer",
        "description": "Game-production mode for compiling prompts into structured requests, runtime-aware blueprints, system packets, continuity artifacts, playable vertical slices, and verification loops.",
        "switchable": True,
    },
    "reverie-ant": {
        "display_name": "Reverie-Ant",
        "description": "Structured long-running execution mode for planning, checkpoints, and verification.",
        "switchable": True,
    },
    "spec-driven": {
        "display_name": "Spec-Driven",
        "description": "Spec authoring mode for requirements, design, and implementation task breakdown.",
        "switchable": True,
    },
    "spec-vibe": {
        "display_name": "Spec-Vibe",
        "description": "Implementation mode for executing approved specs with a lighter workflow.",
        "switchable": True,
    },
    "writer": {
        "display_name": "Writer",
        "description": "Creative writing mode for narrative drafting, continuity, and long-form documentation.",
        "switchable": True,
    },
    "computer-controller": {
        "display_name": "Computer Controller",
        "description": "Pinned NVIDIA desktop-autopilot mode for operating the Windows UI through computer_control.",
        "switchable": False,
        "requires_source": "nvidia",
    },
}


MODE_ALIASES = {
    "reverie": "reverie",
    "default": "reverie",
    "reverie-atlas": "reverie-atlas",
    "atlas": "reverie-atlas",
    "reverie deeper": "reverie-atlas",
    "reverie-deeper": "reverie-atlas",
    "deeper": "reverie-atlas",
    "reverie-gamer": "reverie-gamer",
    "gamer": "reverie-gamer",
    "reverie-spec-driven": "spec-driven",
    "spec-driven": "spec-driven",
    "spec driven": "spec-driven",
    "spec-vibe": "spec-vibe",
    "spec vibe": "spec-vibe",
    "writer": "writer",
    "reverie-ant": "reverie-ant",
    "ant": "reverie-ant",
    "computer-controller": "computer-controller",
    "computer controller": "computer-controller",
    "computer-control": "computer-controller",
    "computer control": "computer-controller",
    "computer": "computer-controller",
}

LEGACY_MODE_ALIASES = {
    # Backward compatibility for historical typo variants.
    "computer-controler": "computer-controller",
    "computer controler": "computer-controller",
}

MODE_ALIASES.update(LEGACY_MODE_ALIASES)


DEFAULT_TOOL_DISCOVERY_PROFILE: Dict[str, tuple[str, ...]] = {
    "focus_categories": tuple(),
    "boost_tools": tuple(),
    "domain_tokens": tuple(),
    "deemphasize_categories": tuple(),
}


MODE_TOOL_DISCOVERY_PROFILES: Dict[str, Dict[str, tuple[str, ...]]] = {
    "reverie": {
        "focus_categories": ("retrieval", "editing", "workspace", "context", "coordination"),
        "boost_tools": (
            "codebase-retrieval",
            "git-commit-retrieval",
            "str_replace_editor",
            "file_ops",
            "command_exec",
            "tool_catalog",
        ),
        "domain_tokens": (
            "bug",
            "build",
            "class",
            "code",
            "file",
            "files",
            "fix",
            "function",
            "repo",
            "repository",
            "refactor",
            "test",
            "verify",
            "workspace",
        ),
        "deemphasize_categories": ("game-design", "game-runtime", "game-modeling", "writer", "atlas", "desktop"),
    },
    "reverie-atlas": {
        "focus_categories": ("atlas", "retrieval", "workspace", "context", "planning"),
        "boost_tools": (
            "atlas_delivery_orchestrator",
            "codebase-retrieval",
            "create_file",
            "command_exec",
            "tool_catalog",
        ),
        "domain_tokens": (
            "appendix",
            "architecture",
            "atlas",
            "charter",
            "contract",
            "delivery",
            "document",
            "documents",
            "handoff",
            "manifest",
            "resume",
            "slice",
            "spec",
            "tracker",
        ),
        "deemphasize_categories": ("game-design", "game-runtime", "game-playtest", "writer", "desktop"),
    },
    "reverie-gamer": {
        "focus_categories": (
            "game-design",
            "game-scaffold",
            "game-runtime",
            "game-playtest",
            "game-data",
            "game-modeling",
            "planning",
            "retrieval",
            "orchestration",
            "image-generation",
        ),
        "boost_tools": (
            "tool_catalog",
            "task_manager",
            "game_design_orchestrator",
            "game_project_scaffolder",
            "reverie_engine",
            "reverie_engine_lite",
            "game_playtest_lab",
            "game_modeling_workbench",
            "game_gdd_manager",
            "game_asset_manager",
            "game_balance_analyzer",
            "game_math_simulator",
            "level_design",
            "story_design",
        ),
        "domain_tokens": (
            "2d",
            "3d",
            "action",
            "asset",
            "balance",
            "blueprint",
            "camera",
            "combat",
            "compiler",
            "controller",
            "economy",
            "engine",
            "first",
            "foundation",
            "gdd",
            "game",
            "godot",
            "hud",
            "level",
            "movement",
            "model",
            "npc",
            "playable",
            "playtest",
            "prototype",
            "production",
            "quest",
            "request",
            "resume",
            "runtime",
            "save",
            "scope",
            "slice",
            "score",
            "expansion",
            "story",
            "system",
            "task",
            "task_graph",
            "telemetry",
            "third",
            "validation",
            "vertical",
            "world",
        ),
        "deemphasize_categories": ("writer", "atlas", "desktop"),
    },
    "reverie-ant": {
        "focus_categories": ("planning", "coordination", "retrieval", "workspace", "context"),
        "boost_tools": (
            "task_boundary",
            "notify_user",
            "task_manager",
            "codebase-retrieval",
            "command_exec",
            "tool_catalog",
        ),
        "domain_tokens": (
            "artifact",
            "checkpoint",
            "phase",
            "plan",
            "planning",
            "progress",
            "review",
            "resume",
            "verification",
            "verify",
        ),
        "deemphasize_categories": ("game-design", "game-runtime", "writer", "desktop"),
    },
    "spec-driven": {
        "focus_categories": ("planning", "retrieval", "editing", "workspace", "context"),
        "boost_tools": (
            "codebase-retrieval",
            "create_file",
            "str_replace_editor",
            "command_exec",
            "tool_catalog",
        ),
        "domain_tokens": (
            "acceptance",
            "architecture",
            "breakdown",
            "design",
            "plan",
            "requirement",
            "spec",
            "task",
        ),
        "deemphasize_categories": ("game-design", "writer", "desktop"),
    },
    "spec-vibe": {
        "focus_categories": ("editing", "workspace", "retrieval", "context"),
        "boost_tools": (
            "codebase-retrieval",
            "str_replace_editor",
            "create_file",
            "command_exec",
            "tool_catalog",
        ),
        "domain_tokens": ("approved", "execute", "implement", "refine", "ship", "spec", "wire"),
        "deemphasize_categories": ("game-design", "writer", "desktop"),
    },
    "writer": {
        "focus_categories": ("writer", "retrieval", "context", "coordination"),
        "boost_tools": (
            "novel_context_manager",
            "consistency_checker",
            "plot_analyzer",
            "ask_clarification",
            "tool_catalog",
        ),
        "domain_tokens": (
            "arc",
            "canon",
            "chapter",
            "character",
            "continuity",
            "dialogue",
            "novel",
            "plot",
            "scene",
            "story",
            "tone",
            "voice",
        ),
        "deemphasize_categories": ("game-design", "game-runtime", "atlas", "desktop"),
    },
    "computer-controller": {
        "focus_categories": ("desktop", "vision", "coordination"),
        "boost_tools": ("computer_control", "vision_upload", "tool_catalog"),
        "domain_tokens": (
            "app",
            "browser",
            "click",
            "cursor",
            "desktop",
            "hotkey",
            "observe",
            "screen",
            "type",
            "ui",
            "window",
        ),
        "deemphasize_categories": ("game-design", "game-runtime", "writer", "atlas", "planning"),
    },
}


def normalize_mode(value: object, default: str = "reverie") -> str:
    """Normalize mode aliases into canonical mode identifiers."""
    raw = str(value or "").strip().lower()
    if not raw:
        return default
    return MODE_ALIASES.get(raw, raw)


def get_mode_metadata(mode: object) -> Dict[str, object]:
    """Return metadata for a mode, falling back to Reverie."""
    return MODE_METADATA.get(normalize_mode(mode), MODE_METADATA["reverie"])


def get_mode_description(mode: object) -> str:
    """Return a short user-facing description for a mode."""
    return str(get_mode_metadata(mode).get("description", "")).strip()


def get_mode_display_name(mode: object) -> str:
    """Return display name for a mode."""
    return str(get_mode_metadata(mode).get("display_name", "Reverie")).strip()


def list_modes(include_computer: bool = True, switchable_only: bool = False) -> List[str]:
    """List supported modes in the preferred display order."""
    result: List[str] = []
    for mode_name, meta in MODE_METADATA.items():
        if not include_computer and mode_name == "computer-controller":
            continue
        if switchable_only and not bool(meta.get("switchable", False)):
            continue
        result.append(mode_name)
    return result


def get_mode_tool_discovery_profile(mode: object) -> Dict[str, tuple[str, ...]]:
    """Return the tool-discovery bias profile for the supplied mode."""
    normalized = normalize_mode(mode)
    profile = MODE_TOOL_DISCOVERY_PROFILES.get(normalized, {})
    return {
        key: tuple(str(item).strip() for item in profile.get(key, DEFAULT_TOOL_DISCOVERY_PROFILE[key]) if str(item).strip())
        for key in DEFAULT_TOOL_DISCOVERY_PROFILE
    }

File: ReverieCli-Rs/crates/reverie-core/src/modes.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Mode {
    Reverie,
    ReverieAtlas,
    ReverieGamer,
    ReverieAnt,
    SpecDriven,
    SpecVibe,
    Writer,
    ComputerController,
}

impl Mode {
    pub fn canonical(self) -> &'static str {
        match self {
            Self::Reverie => "reverie",
            Self::ReverieAtlas => "reverie-atlas",
            Self::ReverieGamer => "reverie-gamer",
            Self::ReverieAnt => "reverie-ant",
            Self::SpecDriven => "spec-driven",
            Self::SpecVibe => "spec-vibe",
            Self::Writer => "writer",
            Self::ComputerController => "computer-controller",
        }
    }

    pub fn display_name(self) -> &'static str {
        match self {
            Self::Reverie => "Reverie",
            Self::ReverieAtlas => "Reverie-Atlas",
            Self::ReverieGamer => "Reverie-Gamer",
            Self::ReverieAnt => "Reverie-Ant",
            Self::SpecDriven => "Spec-Driven",
            Self::SpecVibe => "Spec-Vibe",
            Self::Writer => "Writer",
            Self::ComputerController => "Computer Controller",
        }
    }

    pub fn description(self) -> &'static str {
        match self {
            Self::Reverie => "General coding and automation mode with context retrieval and workspace tools.",
            Self::ReverieAtlas => "Document-driven spec development mode for complex systems.",
            Self::ReverieGamer => "Game-production mode for prompt-to-blueprint and vertical-slice workflows.",
            Self::ReverieAnt => "Structured long-running execution mode for planning, checkpoints, and verification.",
            Self::SpecDriven => "Spec authoring mode for requirements, design, and task breakdown.",
            Self::SpecVibe => "Implementation mode for executing approved specs with a lighter workflow.",
            Self::Writer => "Creative writing mode for narrative drafting and continuity.",
            Self::ComputerController => "Desktop autopilot mode for operating UI surfaces.",
        }
    }

    pub fn switchable(self) -> bool {
        !matches!(self, Self::ComputerController)
    }

    /// Mode-specific system prompt that shapes the LLM's behavior.
    pub fn system_prompt(self) -> &'static str {
        match self {
            Self::Reverie => concat!(
                "You are Reverie, an expert AI coding assistant. ",
                "You help the user with software development tasks including writing, debugging, refactoring, and explaining code. ",
                "You have access to tools for reading/writing files, running commands, searching the codebase, and web search. ",
                "Always prefer editing existing files over creating new ones. ",
                "Use the codebase retrieval tool to understand project structure before making changes. ",
                "When modifying code, make minimal targeted edits using str_replace_editor. ",
                "Explain your reasoning briefly. Be direct and concise. ",
                "Follow existing code style and conventions. Add necessary imports. ",
                "Never fabricate file paths or function names — verify with tools first.",
            ),
            Self::ReverieAtlas => concat!(
                "You are Reverie-Atlas, a document-driven development orchestrator. ",
                "You manage complex multi-phase projects through structured specs, delivery slices, milestones, and blockers. ",
                "Use the atlas_delivery_orchestrator tool to track delivery state, progress, and blockers. ",
                "Break large tasks into ordered delivery slices with clear acceptance criteria. ",
                "Track dependencies between slices and flag blockers proactively. ",
                "Produce checkpoint summaries at natural milestones. ",
                "Maintain a living spec document that evolves as requirements clarify. ",
                "Prefer verified incremental progress over speculative large changes.",
            ),
            Self::ReverieGamer => concat!(
                "You are Reverie-Gamer, a game production assistant. ",
                "You guide prompt-to-blueprint and vertical-slice workflows for game development. ",
                "Use game tools: game_design_orchestrator for pipeline stages, game_gdd_manager for design documents, ",
                "game_playtest_lab for test plans and quality gates, game_asset_manager for asset tracking, ",
                "reverie_engine for engine scaffolding, and story_design/level_design for narrative and world building. ",
                "Follow the production pipeline: design → prototype → vertical slice → playtest → polish. ",
                "Always maintain a GDD as the single source of truth for game systems. ",
                "Track assets, verify quality gates, and score playtest results systematically.",
            ),
            Self::ReverieAnt => concat!(
                "You are Reverie-Ant, a structured execution agent for long-running complex tasks. ",
                "You plan methodically, create checkpoints, and verify each step before advancing. ",
                "Use the task_manager to break work into tracked tasks with clear status. ",
                "Use task_boundary to mark phase transitions and summarize progress. ",
                "Create subagents for parallel independent subtasks when beneficial. ",
                "Always verify your changes compile/pass before marking a step complete. ",
                "If you encounter an unexpected state, pause, report the situation, and ask for guidance rather than guessing. ",
                "Prefer depth-first execution: finish one slice before starting the next.",
            ),
            Self::SpecDriven => concat!(
                "You are Reverie in Spec-Driven mode, focused on requirements analysis and specification authoring. ",
                "Help the user define clear requirements, acceptance criteria, and task breakdowns. ",
                "Produce structured spec documents with sections for goals, constraints, design decisions, and implementation plan. ",
                "Ask clarifying questions when requirements are ambiguous. ",
                "Identify risks, dependencies, and technical constraints early. ",
                "Output specs in Markdown with clear headings, numbered requirements, and decision records. ",
                "Do not implement code in this mode — focus on planning and specification.",
            ),
            Self::SpecVibe => concat!(
                "You are Reverie in Spec-Vibe mode, an implementation-focused assistant that executes approved specifications. ",
                "Work from the existing spec documents to implement changes. ",
                "Follow the spec's implementation plan and acceptance criteria. ",
                "Make incremental changes, test after each step, and report progress. ",
                "If the spec is unclear or needs revision, flag it rather than guessing. ",
                "Use a lighter workflow — less ceremony, more action. ",
                "Prefer small verified commits over large speculative changes.",
            ),
            Self::Writer => concat!(
                "You are Reverie in Writer mode, a creative writing assistant for long-form narrative work. ",
                "Use novel_context_manager to maintain persistent memory of characters, locations, timeline, threads, and chapters. ",
                "Use consistency_checker to validate narrative consistency across the work. ",
                "Use plot_analyzer to evaluate arc progression, pacing, and tension. ",
                "Maintain voice consistency within each piece. Track character development arcs. ",
                "When drafting new content, reference existing context to maintain continuity. ",
                "Flag potential plot holes or timeline inconsistencies proactively. ",
                "Support outlining, drafting, revision, and continuity checking workflows.",
            ),
            Self::ComputerController => concat!(
                "You are Reverie in Computer Controller mode, a desktop automation assistant. ",
                "You can take screenshots, click, type, press keys, scroll, and move the mouse. ",
                "Use the computer_control tool to interact with the user's desktop. ",
                "Always take a screenshot first to understand the current screen state before acting. ",
                "Describe what you see and what you plan to do before executing actions. ",
                "Be cautious — avoid destructive actions without explicit confirmation. ",
                "Chain small precise actions rather than complex sequences. ",
                "If you lose track of the UI state, take another screenshot to re-orient.",
            ),
        }
    }
}

pub fn normalize_mode(value: impl AsRef<str>) -> Mode {
    match value.as_ref().trim().to_ascii_lowercase().as_str() {
        "" | "default" | "reverie" => Mode::Reverie,
        "atlas" | "deeper" | "reverie deeper" | "reverie-deeper" | "reverie-atlas" => {
            Mode::ReverieAtlas
        }
        "gamer" | "reverie-gamer" => Mode::ReverieGamer,
        "ant" | "reverie-ant" => Mode::ReverieAnt,
        "reverie-spec-driven" | "spec driven" | "spec-driven" => Mode::SpecDriven,
        "spec vibe" | "spec-vibe" => Mode::SpecVibe,
        "writer" => Mode::Writer,
        "computer"
        | "computer-control"
        | "computer-controler"
        | "computer controller"
        | "computer controler"
        | "computer-controller" => Mode::ComputerController,
        _ => Mode::Reverie,
    }
}

pub fn list_modes(include_computer: bool, switchable_only: bool) -> Vec<Mode> {
    let modes = vec![
        Mode::Reverie,
        Mode::ReverieAtlas,
        Mode::ReverieGamer,
        Mode::ReverieAnt,
        Mode::SpecDriven,
        Mode::SpecVibe,
        Mode::Writer,
        Mode::ComputerController,
    ];
    modes
        .into_iter()
        .filter(|mode| include_computer || !matches!(mode, Mode::ComputerController))
        .filter(|mode| !switchable_only || mode.switchable())
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn aliases_match_python_surface() {
        assert_eq!(normalize_mode("gamer"), Mode::ReverieGamer);
        assert_eq!(
            normalize_mode("computer controler"),
            Mode::ComputerController
        );
        assert_eq!(normalize_mode("spec vibe"), Mode::SpecVibe);
    }
}
